### PR TITLE
Initial stab at GitHub actions for some simple PR checks

### PR DESCRIPTION
**Pull request by orta**
Creates a GitHub Workflow, which runs on every PR - which uses the [danger JS GitHub Action](https://github.com/marketplace/actions/danger-js-action) to adds the following automation on a PR:

- If someone on the team opens a PR, assign it to themselves
- If someone not on the team opens a PR, if the title or body of the PR doesn't reference an Issue with an assigned Milestone, comment in the PR and add a label for it

---

I had two options for setting up danger. Either add it as an nom dependency, or use the GitHub action. Conceptually the trade of is:

- via GitHub Action: runs danger completely self hosted in a docker container, and doesn't influence the tsc codebase at all

- via `npm run danger`: adds all of danger's dependencies to typescript's dev env (boo) but is probably a minute faster on CI.

Because this isn't going to be the slowest part of the CI process, I figure it's fine to go with the lack of dependencies for a longer run. It doesn't have to be instant-instant. To make that work, I added the GitHub subset of the Danger d.ts file into the repo in the .github folder. The DTS changes very very rarely, and has never broke backwards compatibility.

### CODE REVIEW

**Comment by weswigham** [comment]
@DanielRosenwasser you should tell @orta about fabric bot

**Comment by RyanCavanaugh** [comment]
I'd prefer long-term we get off Fabric bot; now that there are (good?) native tools available we should prefer them for stability, lack of dependence on another team's infrastructure, and for the sake of having our auto rules be publicly viewable/PR-able.

**Comment by weswigham** [comment]
@orta in thaaaaaat case.... Shall we start migrating the rest of the automation tasks, then?

**Comment by RyanCavanaugh** [comment]
No rush, but yeah, eventually. I wanted to run a few Actions here while the Beta's going to see how it goes before going all-in

**Comment by orta** [comment]
Coming back to this, it looks like this isn't _really_ feasible to do for everyone. The access tokens for setting labels will be readonly when it's not from a contributor. Meaning we can't make the API call to add the label.

I've worked around [this before](https://github.com/orta/gh-commentify) in a different context, and in theory we can make a cheap azure function which can have a allowlisted set of labels which this can trigger instead?

**Comment by weswigham** [comment]
> I've worked around this before in a different context, and in theory we can make a cheap azure function

Isn't the github action _essentially_ an azure function itself? What's preventing use from using a more privileged token? Is it just the issue of the code for the action being in the repo itself? If that's so - don't actions always execute their code from `master` branch, and not PRs, so it's not a problem? (And if they don't: can we make it so they can, somehow?)

**Comment by orta** [comment]
Under the hood - yeah probably azure functions. 

The built-in secrets support I'd think of it a lot more like a traditional CI service like Travis, this is [the permission model](https://help.github.com/en/articles/virtual-environments-for-github-actions#token-permissions) for the GitHub Access token _they_ generate per run for you. Basically if you push a branch from the same repo you get write access, otherwise your token is read-only (which can comment).

It doesn't always run master, it runs whatever version is in your current PR - so you could edit and work on it

**Comment by weswigham** [comment]
> Basically if you push a branch from the same repo you get write access, otherwise your token is read-only (which can comment).

That's for the `push` trigger, yeah? But what about the "new issue" (which should include PRs) trigger? That should still be keyed off `master`, no?

**Comment by orta** [comment]
I would hope so, great thinking! I'll refactor this up to do that. 💯 / 💯

File: package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
     },
     "devDependencies": {
         "@octokit/rest": "latest",
+        "@octokit/webhooks": "latest",
```
[the remark this answers is not in the document]
Comment (orta) [reply package.json]: Just keeping the dependency tree small mainly, Danger's types are bundled - so we'd have all of the danger dependencies downloaded to have access. `@octokit/webhooks` on the other hand has only one dep and it's already in the tree.

File: .github/workflows/js/issue.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// extra runtime APIs for working with GitHub.
+
+// Docs for Danger: https://danger.systems/js/
+
+// tslint:disable-next-line: no-implicit-dependencies
+import { danger } from "danger";
+import { WebhookPayloadIssues } from "@octokit/webhooks";
+
+// Danger extracts the webhook and passes it as the first param to the default function
+export default async (webhook: WebhookPayloadIssues) => {
```
[the remark this answers is not in the document]
Comment (orta) [reply .github/workflows/js/issue.ts]: Bah, I did a bad rebase

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (weswigham) [reply package.json]: > Just keeping the dependency tree small mainly

Yeah, I absolutely do not care - none of our users need to install it, just us, and we probably only do it once in awhile. (And CI is CI, it's time doesn't matter until it does) Just add the dev dependency.

[the remark this answers is not in the document]
Comment (orta) [reply package.json]: Works for me 👍

File: package.json [hunk 3]
```diff
@@ -61,6 +62,7 @@
         "chai": "latest",
         "chalk": "latest",
         "convert-source-map": "latest",
+        "danger": "^9.2.0",
```
Comment (weswigham) [inline package.json]: Not `latest`? 🤔  Do we have a reason to fix the version yet?

File: .github/workflows/js/issue.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  const isCollaboratorTeamResponse = await octokit.repos.checkCollaborator({ repo: repo.name, owner: repo.owner.login, username: author.login });
+  const accessToTypeScript = isCollaboratorTeamResponse.data.permission as string;
+
+  // This may be too lax, if so, it can be switched to checking a GitHub team
+  const isTSTeamMember = (accessToTypeScript === "admin" || accessToTypeScript === "write");
+
+  // When there isn't an assignee, assign it to yourself if you're on the team
+  if (issue.assignees.length === 0 && isTSTeamMember) {
```
Comment (weswigham) [inline .github/workflows/js/issue.ts]: `@typescript-bot` has write-level permissions but shouldn't be assigned when it opens a PR - methinks the bot needs an exception here.

**Comment by orta** [comment]
Hrm, so - running the work on issues _could_ work, but I can't confirm it until this is merged and can see the results. Will try on my own repo.

**Comment by orta** [comment]
Looks like the underlaying security model for GH actions just isn't very useful in these cases - https://github.com/actions/labeler/issues/12

Might have to build on top of the existing bot infra

**Comment by orta** [comment]
Closing this out in favor of some serverless functions on azure. 

Because of the security model of Actions, we couldn't make changes like assignments or adding labels to PR which come from forks without doing some pretty funky stuff:

 - use a scheduled task for anything related to pull requests
 - use a normal action for issues

This was because pull requests from forks don't have access to secrets.

This mess was just a bit meh. In response I've cut down most of the useful abstractions in Peril into a small function app in https://github.com/microsoft/TypeScript-repos-automation/pull/1

Here is the equivalent of [this Dangerfile](https://github.com/microsoft/TypeScript-repos-automation/blob/init/src/checks/assignSelfToNewPullRequest.ts) and [it's tests](https://github.com/microsoft/TypeScript-repos-automation/blob/init/src/checks/assignSelfToNewPullRequest.test.ts)

**Pull request closed without merging**